staging/src/k8s.io/apiserver/pkg/server/serve.go

```diff
@@ -224,22 +238,216 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
 	return byName, nil
 }
 
+func (s *ServingInfo) Profile(w http.ResponseWriter, r *http.Request) {
+	encoder := json.NewEncoder(w)
+	encoder.SetIndent("", "\t")
+	err := encoder.Encode(&s.IPLimit)
+	if err != nil {
+		// glog.V(2).Infof("Error marshaling secure server info: %v", err)
+		glog.Infof("Error marshaling secure server info: %v", err)
+	}
+}
+
+func NoSecureHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "{ \"Error\": \"No secure server configured\" }")
+}
+
+// tcpConnectionChainListener is responsible to invoking any connection handling
+// code which is invoked after AcceptTCP() has been called but before the
+// connection is usable by the server for communicating with the client. We
+// assume if an error is returned from a chain handler that it will return a nil
+// conn and have already closed that conn. Init() is used to allow low levels tweaks
+// on the TCPConn.
+type tcpConnectionChainHandler interface {
+	Init(*net.TCPConn) error
+}
+
+type tcpConnectionChainListener struct {
+	*net.TCPListener
+	chain []tcpConnectionChainHandler
+}
+
+func (l tcpConnectionChainListener) Accept() (net.Conn, error) {
+	tc, err := l.AcceptTCP()
+	if err != nil {
+		return nil, err
+	}
+	for _, handler := range l.chain {
+		err = handler.Init(tc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return tc, nil
+}
+
 // tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
 // connections. It's used by ListenAndServe and ListenAndServeTLS so
 // dead TCP connections (e.g. closing laptop mid-download) eventually
 // go away.
 //
 // Copied from Go 1.7.2 net/http/server.go
-type tcpKeepAliveListener struct {
-	*net.TCPListener
+type tcpKeepAliveHandler struct{}
+
+func (ln tcpKeepAliveHandler) Init(conn *net.TCPConn) error {
+	conn.SetKeepAlive(true)
+	conn.SetKeepAlivePeriod(defaultKeepAlivePeriod)
+	return nil
+}
+
+// ipBasedLimitHandler accepts at most n simultaneous connections from a
+// given IP address. We also limit the total simultaneous connections.
+// None of this is applied to localhost/loopback connections.
+//
+// Based on Go 1.7.4 net/netutil/listen.go
+// TODO: (cheftako) : IPv6, CIDR Range rather than individual IP Addresses?
+// Considered using atomic for per IP limit value,
+// but problem initing map value safely
+
```

Comment:
remove blank line (I guess not important here, but for public types a blank line here will break godoc).